src/neo/Ledger/Blockchain.cs

```diff
@@ -221,21 +228,54 @@ private VerifyResult OnNewBlock(Block block)
         {
             DataCache snapshot = View;
             uint currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
+            uint headerHeight = HeaderCache.GetCurrentIndex(snapshot);
             if (block.Index <= currentHeight)
                 return VerifyResult.AlreadyExists;
-            if (block.Index - 1 > currentHeight)
+            if (block.Index - 1 > headerHeight)
             {
                 AddUnverifiedBlockToCache(block);
                 return VerifyResult.UnableToVerify;
             }
-            if (block.Index == currentHeight + 1)
+            if (block.Index == headerHeight + 1)
             {
                 if (!block.Verify(snapshot))
                     return VerifyResult.Invalid;
-                block_cache.TryAdd(block.Hash, block);
-                block_cache_unverified.Remove(block.Index);
-                Persist(block);
-                if (block_cache_unverified.TryGetValue(block.Index + 1, out var unverifiedBlocks))
+            }
+            else
+            {
+                if (!block.Hash.Equals(HeaderCache[block.Index].Hash))
```

Comment:
if it's `Full` it will return null instead of check in the storage?

Reply:
Do you mean `OnNewHeaders`?

Reply:
HeaderCache[block.Index] will not return null. If HeaderCache is full, it will not accept new headers, until oldest headers removed due to block persisting

Reply:
But if don't have the last header and we receive the last block, we should be able to use the block without the header

Reply:
In such cases headercache should be null or only contains the last persisted block's header, so `headerHeight` should be equal to `currentHeight`. Then this block should have been checked in `if (block.Index == headerHeight + 1)` branch